funsor/sum_product.py

```diff
@@ -193,15 +193,41 @@ def Slice(name, *args):
     return Tensor(data, inputs, dtype=bound)
 
 
-def sequential_sum_product(sum_op, prod_op, trans, time, prev, curr):
+def naive_sequential_sum_product(sum_op, prod_op, trans, time, step):
+    assert isinstance(sum_op, AssociativeOp)
+    assert isinstance(prod_op, AssociativeOp)
+    assert isinstance(trans, Funsor)
+    assert isinstance(time, str)
+    assert isinstance(step, dict)
+    assert all(isinstance(k, str) for k in step.keys())
+    assert all(isinstance(v, str) for v in step.values())
+    if time not in trans.inputs:
+        return trans  # edge case of a single time step
+
+    step = OrderedDict(sorted(step.items()))
+    drop = tuple("_drop_{}".format(i) for i in range(len(step)))
+    prev_to_drop = dict(zip(step.keys(), drop))
+    curr_to_drop = dict(zip(step.values(), drop))
+    drop = frozenset(drop)
+
+    factors = [trans(**{time: t}) for t in range(trans.inputs[time].size)]
+    while len(factors) > 1:
+        y = factors.pop()(**prev_to_drop)
+        x = factors.pop()(**curr_to_drop)
+        xy = prod_op(x, y).reduce(sum_op, drop)
+        factors.append(xy)
+    return factors[0]
+
+
+def sequential_sum_product(sum_op, prod_op, trans, time, step):
     """
-    For a funsor ``trans`` with dimensions ``time``, ``prev`` and ``curr``,
-    computes a recursion equivalent to::
+    For a funsor ``trans`` with dimensions ``time``, ``step.keys()``,
+    and ``step.values()``, computes a recursion equivalent to::
 
         tail_time = 1 + arange("time", trans.inputs["time"].size - 1)
         tail = sequential_sum_product(sum_op, prod_op,
                                       trans(time=tail_time),
-                                      "time", "prev", "curr")
+                                      "time", step)
         return prod_op(trans(time=0)(curr="drop"), tail(prev="drop")) \
            .reduce(sum_op, "drop")
```

Comment:
nit: update these docstrings as well

Reply:
I've updated while intentionally restricting to a single variable case, since the multivariable case seems too complex for a docstring.